pkg/master/controller_test.go

```diff
@@ -305,15 +306,53 @@ func TestSetEndpoints(t *testing.T) {
 					ObjectMeta: om("foo"),
 					Subsets: []api.EndpointSubset{{
 						Addresses: []api.EndpointAddress{{IP: "1.2.3.4"}},
-						Ports: []api.EndpointPort{
-							{Name: "bar", Port: 1000, Protocol: "TCP"},
-							{Name: "foo", Port: 8080, Protocol: "TCP"},
-							{Name: "baz", Port: 1010, Protocol: "TCP"},
-						},
+						Ports:     []api.EndpointPort{{Name: "foo", Port: 8080, Protocol: "TCP"}},
 					}},
 				}},
 			},
+			expectUpdate: &api.Endpoints{
+				ObjectMeta: om("foo"),
+				Subsets: []api.EndpointSubset{{
+					Addresses: []api.EndpointAddress{{IP: "1.2.3.4"}},
+					Ports: []api.EndpointPort{
+						{Name: "foo", Port: 8080, Protocol: "TCP"},
+						{Name: "bar", Port: 1000, Protocol: "TCP"},
+					},
+				}},
+			},
 		},
+	}
+	for _, test := range reconcile_tests {
+		master := Controller{MasterCount: test.additionalMasters + 1}
+		registry := &registrytest.EndpointRegistry{
+			Endpoints: test.endpoints,
+		}
+		master.EndpointRegistry = registry
+		err := master.ReconcileEndpoints(test.serviceName, net.ParseIP(test.ip), test.endpointPorts, true)
+		if err != nil {
+			t.Errorf("case %q: unexpected error: %v", test.testName, err)
+		}
+		if test.expectUpdate != nil {
+			if len(registry.Updates) != 1 {
+				t.Errorf("case %q: unexpected updates: %v", test.testName, registry.Updates)
+			} else if e, a := test.expectUpdate, &registry.Updates[0]; !reflect.DeepEqual(e, a) {
+				t.Errorf("case %q: expected update:\n%#v\ngot:\n%#v\n", test.testName, e, a)
+			}
+		}
+		if test.expectUpdate == nil && len(registry.Updates) > 0 {
+			t.Errorf("case %q: no update expected, yet saw: %v", test.testName, registry.Updates)
+		}
+	}
+
+	non_reconcile_tests := []struct {
+		testName          string
+		serviceName       string
+		ip                string
+		endpointPorts     []api.EndpointPort
+		additionalMasters int
+		endpoints         *api.EndpointsList
+		expectUpdate      *api.Endpoints // nil means none expected
+	}{
 		{
```

Comment:
at least one non-reconcile test with an incorrect IP, and one with an incorrect format to demonstrate non-port-reconcile case still updates